### PR TITLE
fix(on-demand): Escape glob patterns meta characters

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR improves the PR that was previously [reverted](https://github.com/getsentry/sentry/commit/1153a9e5fde345669b9ca81b0313f914f2d5eee9) (even though no revert was needed but we figured it out only after the revert that json was escaped because it's actually in the standard). To make this more explicit, this PR adds actual snapshots tests of the json representation and uses raw strings to make the code more readable since there is no need of escaping.

### CODE REVIEW

File: src/sentry/snuba/metrics/extraction.py [hunk 1]
```diff
@@ -1531,6 +1531,29 @@ def _get_satisfactory_threshold_and_metric(project: Project) -> tuple[int, str]:
...
+    """
+    i, n = 0, len(value)
+    escaped = ""
+
+    while i < n:
+        c = value[i]
+        i = i + 1
+
+        if c in "[]{}?":
+            escaped += rf"\{c}"
```
Comment (iambriccardo) [inline src/sentry/snuba/metrics/extraction.py:1550]: Used raw strings to make code more readable.

**Review by obostjancic — APPROVED** [review]

File: tests/sentry/snuba/metrics/test_extraction.py [hunk 2]
```diff
@@ -460,6 +462,40 @@ def test_spec_wildcard() -> None:
...
+        "value": [expected_pattern],
+    }
+
+    # We also validate using Relay's glob implementation to make sure the escaping
+    # is interpreted correctly.
+    assert glob_match(title, expected_pattern, ignorecase=True)
+
+    # We want to validate the json output, to make sure that characters are correctly escaped.
+    metric_spec = spec.to_metric_spec(default_project)
+    insta_snapshot(json.dumps(metric_spec))
```
Comment (iambriccardo) [inline tests/sentry/snuba/metrics/test_extraction.py:496]: Added snapshot test to validate that serialization to json is happening with the correct escaping.

File: src/sentry/snuba/metrics/extraction.py [hunk 3]
```diff
@@ -1531,6 +1531,29 @@ def _get_satisfactory_threshold_and_metric(project: Project) -> tuple[int, str]:
...
+        else:
+            escaped += c
+
+    return escaped
```
Comment (asottile-sentry) [inline src/sentry/snuba/metrics/extraction.py:1554]: this function is likely to be *very* slow -- building a string one (or two) characters at a time tends to not go so well in purelib python

`str.translate` is sorta designed for this so maybe try that out instead?

File: tests/sentry/snuba/metrics/test_extraction.py (hunk 2 above)
Comment (asottile-sentry) [inline tests/sentry/snuba/metrics/test_extraction.py:496]: snapshot tests seem very heavy handed for this -- I would recommend instead unit testing the function you wrote

Comment (iambriccardo) [reply tests/sentry/snuba/metrics/test_extraction.py:496]: Perfect! Will do

**Pull request merged**